// @stekoe/ocl.js/dist/components/parser/OclParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OclParser = void 0;
var Expressions = require("../expressions");
var Utils_1 = require("../Utils");
var parser = require("../../../generator/Parser");
parser.yy = {
    Expression: Expressions,
    Utils: Utils_1.Utils
};
var OclParser = /** @class */ (function () {
    function OclParser() {
    }
    OclParser.parseQuery = function (oclExpression, registeredTypes) {
        parser.yy.registeredTypes = registeredTypes || {};
        return parser.parse(oclExpression);
    };
    OclParser.parse = function (oclExpression, labels, registeredTypes) {
        if (labels === void 0) { labels = []; }
        var packageDeclaration = OclParser.parseQuery(oclExpression, registeredTypes);
        packageDeclaration.setExecutionLabels(labels);
        packageDeclaration.setRawOclExpression(oclExpression);
        return packageDeclaration;
    };
    OclParser.registeredTypes = {
        Array: Array,
        Boolean: Boolean,
        Function: Function,
        Number: Number,
        Object: Object,
        String: String
    };
    OclParser.registeredEnums = {};
    return OclParser;
}());
exports.OclParser = OclParser;
